### ISSUE TITLE
suppress "Pattern not found" when a search yields no results

### ISSUE THREAD

**Issue by no-more-secrets** [neovim/neovim#24059]
### Problem

When cmdheight=0, a search (e.g. either `/` or `n`) that yields no results will trigger a "Pattern not found" and then a "Press ENTER or..." in the message box, which is kind of annoying.  Thus I think we need an option to suppress "Pattern not found" when a search yields no results.  I have highlighting of matches turned on, so I know when a search yields results visually, and thus I also know when it does not; I don't need an additional message telling me so in the message box anyway.

I did some searching and vim does not seem to want to implement this (see https://github.com/vim/vim/issues/11530).  An apparently the `shortmess+=c` does not apply to regular searches.

### Expected behavior

When the "Pattern not found" message is disabled then searching for something that yields no results should cause no visual changes.

**Comment by no-more-secrets** [comment]
Here's what I'm doing as a workaround:
```lua
local function norm_unmapped( c )
  return vim.cmd{ cmd='norm', args={c}, bang=true }
end

vim.keymap.set( 'n', 'n', function() pcall( norm_unmapped, 'n' ) end )
vim.keymap.set( 'n', 'N', function() pcall( norm_unmapped, 'N' ) end )
vim.cmd[[nnoremap / :silent! /]]
```
but this seems kind of hacky, and it may well cause side effects that I'm not aware of yet.  Btw, how would one do that last line in pure Lua?  Would be nicer just to have an option to disable that particular error to begin with.

**Comment by no-more-secrets** [comment]
Another option might be that the error message continues to show, but when cmdheight=0, it just shows temporarily then goes away (I think there are other cases where that happens).

**Comment by justinmk** [comment]
`shortmess=s` could be extended to do this, would need to see if there are objections.

> When cmdheight=0, a search (e.g. either `/` or `n`) that yields no results will trigger a "Pattern not found" and then a "Press ENTER or..." in the message box, which is kind of annoying.

That is something that needs to be solved by `cmdheight=0` in general. Not by various workarounds.

Probably the implementation of vim's `:echowindow` (currently missing in Nvim) could be re-used to ensure that press-enter prompt is never shown for `cmdheight=0`.

**Comment by no-more-secrets** [comment]
Turns out that my workaround above (for replacing the `/` command in particular):
```lua
vim.cmd[[nnoremap / :silent! /]]
```
breaks incremental highlighting of search results.  In order to fix that I have implemented...
```lua
-- This monstrosity manually recreates the behavior of the de-
-- fault / search command, with regard to incremental highlight-
-- ing, cursor placement, and behavior upon cancelling out of the
-- search before hitting enter.
--
-- It does this all this for the purpose of avoiding the error
-- message that normally results when the search yields no
-- results, which doesn't work well with cmdheight=0.
local function search_cmd_with_manual_incsearch_highlight()
  local function set_search( what )
    vim.fn.setreg( '/', what )
    vim.o.hls = true
    vim.cmd[[redraw!]]
  end
  local GROUP = 'ManualIncSearch'
  local old_search = vim.fn.getreg( '/' )
  local old_v_hlsearch = vim.v.hlsearch
  vim.api.nvim_create_autocmd( 'CmdLineChanged', {
    group=vim.api.nvim_create_augroup( GROUP, { clear=true } ),
    callback=function() set_search( vim.fn.getcmdline() ) end,
  } )
  local input = vim.fn.input( '/' )
  vim.api.nvim_del_augroup_by_name( GROUP )
  if #input > 0 then
    set_search( input )
    vim.cmd.normal( 'n' )
    return
  end
  -- The search was cancelled, so try to restore the state to
  -- what it was previously, both with regard to the search
  -- phrase and whether highlighting was visible or not.
  vim.fn.setreg( '/', old_search )
  if old_v_hlsearch == 0 then
    vim.cmd.nohlsearch()
  else
    vim.o.hls = true
  end
  vim.cmd[[redraw!]]
end

vim.keymap.set( 'n', '/', search_cmd_with_manual_incsearch_highlight, { silent=true } )
```
which is kind of unfortunate... perhaps the author of `cmdheight=0` could take a look at fixing this so that we don't need workarounds such as the one above.

**Comment by justinmk** [comment]
> > When cmdheight=0, a search (e.g. either `/` or `n`) that yields no results will trigger a "Pattern not found" and then a "Press ENTER or..." in the message box, which is kind of annoying.
> 
> That is something that needs to be solved by `cmdheight=0` in general. Not by various workarounds.

This is fixed by `:help ui2` in Nvim 0.12
